packages/core/src/validate.js

```diff
@@ -300,17 +301,27 @@ export function withIdRefPrefix(schemaNode) {
  */
 export function isValid(schema, data, rootSchema) {
   try {
-    // add the rootSchema ROOT_SCHEMA_PREFIX as id.
+    // add the rootSchema. if it has no $id, use ROOT_SCHEMA_PREFIX as the cache key.
     // then rewrite the schema ref's to point to the rootSchema
     // this accounts for the case where schema have references to models
     // that lives in the rootSchema but not in the schema in question.
-    return ajv
-      .addSchema(rootSchema, ROOT_SCHEMA_PREFIX)
-      .validate(withIdRefPrefix(schema), data);
+    let rootSchemaAdded;
+    if (rootSchema.$id) {
+      rootSchemaAdded = !!ajv.getSchema(rootSchema.$id);
+    } else {
+      rootSchemaAdded = !!ajv.getSchema(ROOT_SCHEMA_PREFIX);
+    }
+
+    if (!rootSchemaAdded) {
+      if (rootSchema.$id) {
+        ajv.addSchema(rootSchema);
+      } else {
+        ajv.addSchema(rootSchema, ROOT_SCHEMA_PREFIX);
+      }
+    }
+    return ajv.validate(withIdRefPrefix(schema), data);
   } catch (e) {
+    console.warn("Encountered error while validating schema", e);
     return false;
-  } finally {
-    // make sure we remove the rootSchema from the global ajv instance
-    ajv.removeSchema(ROOT_SCHEMA_PREFIX);
```

Comment:
Why is the root schema being removed (before my change)?

Since we now need to run the validator frequently to evaluate conditional subschemas, keeping the root schema in the cache is essential for performance, especially for large schemas.

Reply:
I think I've figured it out: this function uses the same ajv instance as `validateFormData`, which has some advantages and some drawbacks. I think with this new implementation (not removing the root schema), validation breaks for certain schemas. I'm not happy with the performance of the current implementation, though, so I'll have to figure something else out.

Reply:
One cause of performance issues here is that we can't take advantage of AJV's cache because `withIdRefPrefix` creates a new object every time. We see a substantial improvement in performance if we memoize `withIdRefPrefix`, because `validate` sees the same schema and can use the ajv cache instead of recompiling. I'm not yet sure that memoization is 100% safe here, but it seems like a possible path 


Another thing I've noticed: we get different behavior with the current implementation if the root schema has an $id. When an $id is present, I _think_ AJV totally ignores the root schema prefix when we call `addSchema(rootSchema, ROOT_SCHEMA_PREFIX)`, so the `removeSchema` call never works. Not 100% on this either, though.

Reply:
As for `$id` issues, I've filed #2471, ideally would like to address it here (at the risk of increasing scope of this PR) since how RJSF uses AJV is closely related with adding conditional support